tokio/src/sync/semaphore.rs

```diff
@@ -86,21 +86,33 @@ impl Semaphore {
         self.ll_sem.release(n);
     }
 
-    /// Acquires permit from the semaphore.
+    /// Acquires one permit from the semaphore.
+    #[inline]
     pub async fn acquire(&self) -> SemaphorePermit<'_> {
-        self.ll_sem.acquire(1).await.unwrap();
+        self.acquire_n(1).await
+    }
+
+    /// Acquires a number of permits from the semaphore.
+    pub async fn acquire_n(&self, num_permits: u16) -> SemaphorePermit<'_> {
+        self.ll_sem.acquire(num_permits).await.unwrap();
         SemaphorePermit {
             sem: &self,
-            permits: 1,
+            permits: num_permits,
         }
```

Comment:
Yes, that sounds interesting. For me for another reason: `u16` is too small. I want to use multiple permits to limit peak memory usage, but `u16` requires this tracking to be quite coarse.

Reply:
Acquiring more than `u16::max` permits in one operation would be a bigger change, since you'd have to propagate it through the underlying low-level semaphore implementation as well.

Reply:
I've made  #2607 for this.